### DISCUSSION TITLE
guides/preprocessing_annotated_data/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#17639] [Docs]
# guides/preprocessing_annotated_data/

Learn essential data preprocessing techniques for annotated computer vision data, including resizing, normalizing, augmenting, and splitting datasets for optimal model training.

https://docs.ultralytics.com/guides/preprocessing_annotated_data/

**Comment by astronitya** [comment]
Hello, 
I have a question related to training image size, I work with astronomical images, and the image size I have are in size of 2kx4k, can I train a custom model with this image size?

**Reply by glenn-jocher** [reply]
@astronitya yes, you can train a custom model with 2kx4k images using YOLO11. The 'imgsz' parameter allows you to specify input sizes, and the model will resize images accordingly while maintaining the aspect ratio. For optimal performance, consider resizing images to a consistent size that suits your computational resources. More details can be found in the [Data Preprocessing Techniques guide](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

**Reply by astronitya** [reply]
What do you exactly mean by "the model will resize images accordingly while maintaining the aspect ratio." Will it train for the whole 2kx4k image, or just a part of it?

**Reply by glenn-jocher** [reply]
The model will resize the entire 2kx4k image to the specified size, such as 640, while maintaining the aspect ratio, meaning it scales the largest dimension to 640 pixels and adjusts the other dimension proportionally. This ensures the whole image is used during training, not just a part of it.

**Reply by astronitya** [reply]
Thankyou for the response glenn. I tried to train my whole image. The issue here is that I have very minor details I have to train my model for. Resizing seems to lose those details.

**Reply by glenn-jocher** [reply]
If resizing is causing a loss of detail in your astronomical images, consider using image tiling to divide your larger images into smaller segments without reducing resolution. This way, you can maintain detail and train the model on these high-resolution fragments. Adjust your dataset configuration to accommodate this approach. For more details, refer to the [Data Preprocessing Techniques guide](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

**Reply by valschmidt** [reply]
@glenn-jocher, could you comment more on best practices for both training and inference with regard to image size? We did a test recently in which we evaluated the performance of a model with the `imgsz` parameter set to varying values, both larger and smaller than the native image size. We found that the model performs much worse (almost by half) when `imgsz` is not matched to the actual size of the image. We thought about tiling too, as you have suggested, but this comes at a computational price and our application is real-time. This suggests that to properly evaluate a model, and indeed deploy it, the value of `imgsz` should not be set to a constant value, but should be automatically varied to match the input images. Is this currently possible with any of the CLI or canned tools provided by ultralytics?

The sensitivity of the performance of the model to the `imgsz` value during inference then gives me pause for thought regarding how it should be set during training. It seems it too should match the native image size, and anything else is unnecessarily throwing away information. But is it possible using the ultralytics tools to change it on the fly during training? Do I understand this correctly? I'd be curious to hear your recommendations here on the `imgsz` parameter during training.

I guess I am surprised that, given the sensitivity of model performance to this parameter, it is not more widely discussed nor central to the discussions of model performance. I feel like I am missing something fundamental.

**Reply by glenn-jocher** [reply]
Thank you for your detailed observations. Currently, YOLO's `imgsz` parameter does not dynamically adjust to match varying input sizes during training or inference. However, you can explore image tiling for detailed features or use the `rect` parameter for inference to handle non-uniform image dimensions more efficiently. For training, setting `imgsz` close to the native resolution can preserve more detail, but it must balance with computational constraints. For more on preprocessing and resizing techniques, refer to [Data Preprocessing Techniques](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

**Reply by valschmidt** [reply]
Thanks for your reply. Fascinating. 

I am interested to learn more about the change in Yolo from early versions in which the model input was fixed and later versions in which the `imgsz` parameter allowed the model input size to change dynamically. I've tried to look up the seminal papers that introduced each of the versions but I can't seem to find a discussion that directly addresses the development that allowed for a model that can adapt to the image size. 

Would you happen to know in what version of Yolo the `imgsz` parameter was introduced? 
Could you recommend a paper that discusses this major change in the way Yolo is architected?

**Reply by glenn-jocher** [reply]
The introduction of the `imgsz` parameter and dynamic input resizing was first seen in YOLOv5, developed by Ultralytics. This feature wasn't explicitly discussed in a single seminal paper but emerged as part of the iterative advancements in YOLO's architecture. For a historical perspective, you can explore the YOLOv5 repository and its [documentation](https://github.com/ultralytics/ultralytics).

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for your interest in Ultralytics 🚀! It looks like you're diving into data preprocessing techniques for annotated computer vision data. That's a great step for optimizing your model training! Make sure to explore the [guides](https://docs.ultralytics.com/guides/preprocessing_annotated_data/) as they provide essential information on resizing, normalizing, augmenting, and splitting datasets.

If you encounter any issues or have questions while following the guide, feel free to provide additional context or details. An Ultralytics engineer will be along shortly to assist you further.

If this is a 🐛 Bug Report related to the documentation, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/), as it greatly aids in debugging the problem.

For real-time chat and quick questions, join us on [Discord](https://ultralytics.com/discord) 🎧. Alternatively, participate in in-depth discussions on [Discourse](https://community.ultralytics.com) or share your thoughts on our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Before diving into any resolution steps, ensure you're using the latest version of the `ultralytics` package along with its [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in an environment with [**Python>=3.8**](https://www.python.org/) and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Consider running YOLO in one of the following verified environments with pre-installed dependencies:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. Refer to our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. Follow our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Check this badge to verify if all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing.

Happy coding! 🎉

**Comment by IruhDegSet** [comment]
hi, 
I have an issue with my dataset which is imbalanced, I tried many techniques like data augmentaion but the problem is that
with object detection you can't really oversample one classe without affecting other classes in the same image
for example face and eyes detection, the face have two eyes, so the nuber of instances of eyes will be always as twice as the number of faces, even if you try augmentation you still get the same imbalance dataset. 
how to Deal with that, if any one has an idea please share it with me??

**Reply by glenn-jocher** [reply]
@IruhDegSet balancing inherently dependent classes like faces and eyes in object detection is challenging. A potential approach is to use class weighting during training to emphasize underrepresented classes or adjust loss functions to account for class importance. For further insights, explore [data preprocessing techniques](https://docs.ultralytics.com/guides/preprocessing_annotated_data/). Let us know if you need more details!

**Comment by Fatima-khanjar** [comment]
@glenn-jocher Hello,

I hope you are doing well.

I have a question regarding my object detection dataset. I’m facing an issue with class imbalance. As a potential solution, I’m considering applying data augmentation, but I would like to apply it only to specific underrepresented classes, rather than to the entire dataset.

Could you please advise on how to implement class-specific augmentation techniques when working with YOLO?

Best regards,

**Comment by VeeranjaneyuluToka** [comment]
Hi, wanted to visualize the dataset from ultralytics dataloader (before it passed to the model or architecture), is there any samples for the same? wanted to visualize both data and augmented data with labels on them.

**Comment by AlexShaGH** [comment]
Greetings, I’m seeking clarification on image normalization embedded into Ultralytics YOLO11 models and API. I’ve seen in several posts that pixel values brought into [0,1] range within the model or by API by simply dividing values by 255. What if the imagery has 16 bits depth? Or any other than 8 number of bits? Is it taken into account automatically (i.e. divided by 65535 for 16 bits, etc.) while training/inferring or is it left to a practitioner to do extra steps before feeding imagery into the process? Please, advise.

**Comment by Sudharsanam-1** [comment]
Hi , I have a  question related to annotation .
  Which is  optimal  for chinese symbol  annotation? segmentation  or detection .
  Kindly , suggest some tips for model training.

**Reply by glenn-jocher** [reply]
Hi @Sudharsanam-1 — for Chinese symbols, **detection** is usually the best starting point if a bounding box is enough (find where the symbol is); choose **segmentation** only if you truly need pixel-accurate outlines (e.g., stroke-level defects or tight shape measurements), see the [Detection task docs](https://docs.ultralytics.com/tasks/detect/) and [Segmentation task docs](https://docs.ultralytics.com/tasks/segment/). For training tips: keep annotations consistent, split train/val/test *before* any augmentation to avoid leakage, and for small symbols consider increasing `imgsz` (and/or tiling your images) so characters aren’t downscaled away; the [preprocessing guide](https://docs.ultralytics.com/guides/preprocessing_annotated_data/) and [YOLO data augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/) cover the recommended workflow—feel free to share a couple sample images + your label format here and we can suggest a more specific setup.
